lib/kino/control.ex

```diff
@@ -230,33 +232,169 @@ defmodule Kino.Control do
-    Kino.SubscriptionManager.stream(control.attrs.ref)
+  @spec stream(source | list(source)) :: Enumerable.t()
+        when source: t() | Kino.Input.t() | interval()
+  def stream(sources) when is_list(sources) do
+    for source <- sources, do: assert_stream_source!(source)
+
+    tagged_topics = for %{attrs: %{ref: ref}} <- sources, do: {nil, ref}
+    tagged_intervals = for {:interval, ms} <- sources, do: {nil, ms}
+
+    build_stream(tagged_topics, tagged_intervals, fn nil, event -> event end)
+  end
+
+  def stream(source) do
+    stream([source])
+  end
+
+  @doc """
+  Same as `stream/1`, but attaches custom tag to every stream item.
+
+  ## Example
+
+      button = Kino.Control.button("Hello")
+      input = Kino.Input.checkbox("Check")
+
+      for event <- Kino.Control.stream([hello: button, check: input]) do
+        IO.inspect(event)
+      end
+      #=> {:hello, %{origin: #PID<10895.9854.0>, type: :click}}
+      #=> {:check, %{origin: #PID<10895.9854.0>, type: :change, value: true}}
+  """
+  def tagged_stream(entries) when is_list(entries) do
+    for entry <- entries do
+      case entry do
+        {tag, source} when is_atom(tag) ->
+          assert_stream_source!(source)
+
+        _other ->
+          raise ArgumentError, "expected a keyword list, got: #{inspect(entries)}"
+      end
+    end
+
+    tagged_topics = for {tag, %{attrs: %{ref: ref}}} <- entries, do: {tag, ref}
+    tagged_intervals = for {tag, {:interval, ms}} <- entries, do: {tag, ms}
+
+    build_stream(tagged_topics, tagged_intervals, fn tag, event -> {tag, event} end)
+  end
+
+  defp assert_stream_source!(%Kino.Control{}), do: :ok
+  defp assert_stream_source!(%Kino.Input{}), do: :ok
+  defp assert_stream_source!({:interval, ms}) when is_number(ms) and ms > 0, do: :ok
+
+  defp assert_stream_source!(item) do
+    raise ArgumentError,
+          "expected source to be either %Kino.Control{}, %Kino.Input{} or {:interval, ms}, got: #{inspect(item)}"
+  end
+
+  defp build_stream(tagged_topics, tagged_intervals, mapper) do
+    Stream.resource(
+      fn ->
+        ref = make_ref()
+
+        for {tag, topic} <- tagged_topics do
+          Kino.SubscriptionManager.subscribe(topic, self(), {ref, tag}, notify_clear: true)
+        end
+
+        for {tag, ms} <- tagged_intervals do
+          Process.send_after(self(), {{ref, tag}, :__interval__, ms, 0}, ms)
+        end
+
+        topics = Enum.map(tagged_topics, &elem(&1, 1))
+
+        {ref, topics}
+      end,
+      fn {ref, topics} ->
+        receive do
+          {{^ref, tag}, event} ->
+            {[mapper.(tag, event)], {ref, topics}}
+
+          {{^ref, _tag}, :topic_cleared, topic} ->
+            case topics -- [topic] do
+              [] -> {:halt, {ref, []}}
+              topics -> {[], {ref, topics}}
+            end
+
+          {{^ref, tag}, :__interval__, ms, n} ->
+            Process.send_after(self(), {{ref, tag}, :__interval__, ms, n + 1}, ms)
```

Comment:
I think the events emitted should be `%{type: :interval, iteration: n}` or something like that. Please make sure to document the type on `Kino.Control.interval`. :)

Reply:
I was keeping it in line with `Stream.interval/1`, but sounds better to align the events instead yeah!